### PR TITLE
[new release] terminus, terminus-hlc and terminus-cohttp (0.1.0)

### PR DESCRIPTION
**Pull request by maiste**
A generic client to interact with Rest API

- Project page: <a href="https://github.com/maiste/terminus">https://github.com/maiste/terminus</a>
- Documentation: <a href="maiste.github.io/terminus">maiste.github.io/terminus</a>

### Added

- Provide a `Terminus` interface (@maiste)
- Introduce the mock module (@maiste, @art-w)
- Implement to interfaces for `Terminus`, `terminus-hlc` and  `terminus-cohttp` (@maiste)

### CODE REVIEW

**Comment by kit-ty-kate** [comment]
There seem to be many missing dependencies such as:
```
#=== ERROR while compiling terminus-cohttp.0.1.0 ==============================#
# context              2.1.2 | linux/x86_64 | ocaml-base-compiler.4.03.0 | pinned(https://github.com/maiste/terminus/releases/download/0.1.0/terminus-0.1.0.tbz)
# path                 ~/.opam/4.03/.opam-switch/build/terminus-cohttp.0.1.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p terminus-cohttp -j 31 @install
# exit-code            1
# env-file             ~/.opam/log/terminus-cohttp-10439-bdaeac.env
# output-file          ~/.opam/log/terminus-cohttp-10439-bdaeac.out
### output ###
# File "src/dune", line 17, characters 12-20:
# 17 |  (libraries terminus cohttp-lwt-unix)
#                  ^^^^^^^^
# Error: Library "terminus" not found.
# -> required by library "terminus-cohttp" in _build/default/src
# -> required by _build/default/META.terminus-cohttp
# -> required by _build/install/default/lib/terminus-cohttp/META
# -> required by _build/default/terminus-cohttp.install
# -> required by alias install

- File "src/dune", line 17, characters 12-20:
- 17 |  (libraries terminus cohttp-lwt-unix)
-                  ^^^^^^^^
- Error: Library "terminus" not found.
- -> required by library "terminus-cohttp" in _build/default/src
- -> required by _build/default/META.terminus-cohttp
- -> required by _build/install/default/lib/terminus-cohttp/META
- -> required by _build/default/terminus-cohttp.install
- -> required by alias install
```

**Comment by maiste** [comment]
Sorry, my mistake! As everything was in the same repository, I forgot to add the internal dependency from `terminus-hlc` and  `terminus-cohttp` to `terminus`. It should be fixed! 
Thanks :smile:

**Comment by kit-ty-kate** [comment]
For some reason terminus-hlc and cohttp’s tests fail on macOS:
```
#=== ERROR while compiling terminus-hlc.0.1.0 =================================#
# context              2.1.2 | macos/x86_64 | ocaml.4.14.0 | pinned(https://github.com/maiste/terminus/releases/download/0.1.0/terminus-0.1.0.tbz)
# path                 ~/.opam/default/.opam-switch/build/terminus-hlc.0.1.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p terminus-hlc -j 11 @install @runtest
# exit-code            1
# env-file             ~/.opam/log/terminus-hlc-19221-6646f7.env
# output-file          ~/.opam/log/terminus-hlc-19221-6646f7.out
### output ###
# (cd _build/default && /Users/mac1000/.opam/default/bin/ocamlopt.opt -w -40 -g -o tests/hlc_test.exe /Users/mac1000/.opam/default/lib/result/result.cmxa /Users/mac1000/.opam/default/lib/ppx_deriving/runtime/ppx_deriving_runtime.cmxa /Users/mac1000/.opam/default/lib/terminus/terminus.cmxa /Users/mac1000/.opam/default/lib/bigstringaf/bigstringaf.cmxa -I /Users/mac1000/.opam/default/lib/bigstringaf /Users/mac1000/.opam/default/lib/faraday/faraday.cmxa -I /Users/mac1000/.opam/default/lib/ocaml /Users/mac1000/.opam/default/lib/lwt/lwt.cmxa /Users/mac1000/.opam/default/lib/ocaml/unix.cmxa -I /Users/mac1000/.opam/default/lib/ocaml /Users/mac1000/.opam/default/lib/ocaml/bigarray.cmxa -I /Users/mac1000/.opam/default/lib/ocaml /Users/mac1000/.opam/default/lib/ocplib-endian/ocplib_endian.cmxa /Users/mac1000/.opam/default/lib/ocplib-endian/bigstring/ocplib_endian_bigstring.cmxa /Users/mac1000/.opam/default/lib/ocaml/threads/threads.cmxa -I /Users/mac1000/.opam/default/lib/ocaml /Users/mac1000/.opam/default/lib/lwt/unix/lwt_unix.cmxa -I /Users/mac1000/.opam/default/lib/lwt/unix /Users/mac1000/.opam/default/lib/faraday-lwt/faraday_lwt.cmxa /Users/mac1000/.opam/default/lib/faraday-lwt-unix/faraday_lwt_unix.cmxa /Users/mac1000/.opam/default/lib/angstrom/angstrom.cmxa /Users/mac1000/.opam/default/lib/base64/base64.cmxa /Users/mac1000/.opam/default/lib/httpaf/httpaf.cmxa /Users/mac1000/.opam/default/lib/psq/psq.cmxa /Users/mac1000/.opam/default/lib/hpack/hpack.cmxa /Users/mac1000/.opam/default/lib/h2/h2.cmxa /Users/mac1000/.opam/default/lib/cstruct/cstruct.cmxa -I /Users/mac1000/.opam/default/lib/cstruct /Users/mac1000/.opam/default/lib/sexplib0/sexplib0.cmxa /Users/mac1000/.opam/default/lib/base/caml/caml.cmxa /Users/mac1000/.opam/default/lib/parsexp/parsexp.cmxa /Users/mac1000/.opam/default/lib/sexplib/sexplib.cmxa /Users/mac1000/.opam/default/lib/cstruct-sexp/cstruct_sexp.cmxa /Users/mac1000/.opam/default/lib/logs/logs.cmxa /Users/mac1000/.opam/default/lib/eqaf/eqaf.cmxa /Users/mac1000/.opam/default/lib/eqaf/bigstring/eqaf_bigstring.cmxa /Users/mac1000/.opam/default/lib/eqaf/cstruct/eqaf_cstruct.cmxa /Users/mac1000/.opam/default/lib/mirage-crypto/mirage_crypto.cmxa -I /Users/mac1000/.opam/default/lib/mirage-crypto /Users/mac1000/.opam/default/lib/hkdf/hkdf.cmxa /Users/mac1000/.opam/default/lib/mirage-crypto-rng/mirage_crypto_rng.cmxa /Users/mac1000/.opam/default/lib/zarith/zarith.cmxa -I /Users/mac1000/.opam/default/lib/zarith /Users/mac1000/.opam/default/lib/mirage-crypto-pk/mirage_crypto_pk.cmxa /Users/mac1000/.opam/default/lib/ptime/ptime.cmxa /Users/mac1000/.opam/default/lib/asn1-combinators/asn1_combinators.cmxa /Users/mac1000/.opam/default/lib/fmt/fmt.cmxa /Users/mac1000/.opam/default/lib/gmap/gmap.cmxa /Users/mac1000/.opam/default/lib/domain-name/domain_name.cmxa /Users/mac1000/.opam/default/lib/mirage-crypto-ec/mirage_crypto_ec.cmxa -I /Users/mac1000/.opam/default/lib/mirage-crypto-ec /Users/mac1000/.opam/default/lib/pbkdf/pbkdf.cmxa /Users/mac1000/.opam/default/lib/macaddr/macaddr.cmxa /Users/mac1000/.opam/default/lib/ipaddr/ipaddr.cmxa /Users/mac1000/.opam/default/lib/x509/x509.cmxa /Users/mac1000/.opam/default/lib/ppx_sexp_conv/runtime-lib/ppx_sexp_conv_lib.cmxa /Users/mac1000/.opam/default/lib/ipaddr-sexp/ipaddr_sexp.cmxa /Users/mac1000/.opam/default/lib/stdlib-shims/stdlib_shims.cmxa /Users/mac1000/.opam/default/lib/tls/tls.cmxa /Users/mac1000/.opam/default/lib/ptime/clock/os/ptime_clock.cmxa -I /Users/mac1000/.opam/default/lib/ptime/clock/os /Users/mac1000/.opam/default/lib/mirage-crypto-rng/unix/mirage_crypto_rng_unix.cmxa -I /Users/mac1000/.opam/default/lib/mirage-crypto-rng/unix /Users/mac1000/.opam/default/lib/duration/duration.cmxa /Users/mac1000/.opam/default/lib/mtime/mtime.cmxa /Users/mac1000/.opam/default/lib/mtime/clock/os/mtime_clock.cmxa -I /Users/mac1000/.opam/default/lib/mtime/clock/os /Users/mac1000/.opam/default/lib/mirage-crypto-rng/lwt/mirage_crypto_rng_lwt.cmxa /Users/mac1000/.opam/default/lib/tls/lwt/tls_lwt.cmxa /Users/mac1000/.opam/default/lib/astring/astring.cmxa /Users/mac1000/.opam/default/lib/rresult/rresult.cmxa /Users/mac1000/.opam/default/lib/fpath/fpath.cmxa /Users/mac1000/.opam/default/lib/bos/bos.cmxa /Users/mac1000/.opam/default/lib/ca-certs/ca_certs.cmxa -I /Users/mac1000/.opam/default/lib/ca-certs /Users/mac1000/.opam/default/lib/metrics/metrics.cmxa /Users/mac1000/.opam/default/lib/dns/dns.cmxa /Users/mac1000/.opam/default/lib/lru/lru.cmxa /Users/mac1000/.opam/default/lib/dns/cache/dns_cache.cmxa /Users/mac1000/.opam/default/lib/randomconv/randomconv.cmxa /Users/mac1000/.opam/default/lib/dns-client/dns_client.cmxa /Users/mac1000/.opam/default/lib/dns-client/resolvconf/dns_resolvconv.cmxa /Users/mac1000/.opam/default/lib/ipaddr/unix/ipaddr_unix.cmxa /Users/mac1000/.opam/default/lib/happy-eyeballs/happy_eyeballs.cmxa /Users/mac1000/.opam/default/lib/dns-client/lwt/dns_client_lwt.cmxa /Users/mac1000/.opam/default/lib/happy-eyeballs-lwt/happy_eyeballs_lwt.cmxa /Users/mac1000/.opam/default/lib/http-lwt-client/http_lwt_client.cmxa src/terminus_hlc.cmxa /Users/mac1000/.opam/default/lib/cmdliner/cmdliner.cmxa /Users/mac1000/.opam/default/lib/uutf/uutf.cmxa /Users/mac1000/.opam/default/lib/alcotest/stdlib_ext/alcotest_stdlib_ext.cmxa /Users/mac1000/.opam/default/lib/fmt/fmt_cli.cmxa /Users/mac1000/.opam/default/lib/re/re.cmxa /Users/mac1000/.opam/default/lib/alcotest/engine/alcotest_engine.cmxa /Users/mac1000/.opam/default/lib/fmt/fmt_tty.cmxa /Users/mac1000/.opam/default/lib/alcotest/alcotest.cmxa -I /Users/mac1000/.opam/default/lib/alcotest /Users/mac1000/.opam/default/lib/alcotest-lwt/alcotest_lwt.cmxa /Users/mac1000/.opam/default/lib/jsonm/jsonm.cmxa /Users/mac1000/.opam/default/lib/hex/hex.cmxa /Users/mac1000/.opam/default/lib/ezjsonm/ezjsonm.cmxa /Users/mac1000/.opam/default/lib/httpaf-lwt-unix/httpaf_lwt_unix.cmxa tests/helpers/helpers.cmxa tests/.hlc_test.eobjs/native/dune__exe__Hlc_test.cmx)
# ld: warning: directory not found for option '-L/opt/local/lib'
# File "tests/dune", line 3, characters 7-15:
# 3 |  (name hlc_test)
#            ^^^^^^^^
# (cd _build/default/tests && ./hlc_test.exe)
# Testing `Http Http-lwt-client module'.
# This run has ID `PLG09SFO'.
# 
# > [FAIL]        call          0   GET Method.
#   [FAIL]        call          1   POST Method.
#   [FAIL]        call          2   PUT Method.
#   [FAIL]        call          3   DELETE Method.
# 
# ┌──────────────────────────────────────────────────────────────────────────────┐
# │ [FAIL]        call          0   GET Method.                                  │
# └──────────────────────────────────────────────────────────────────────────────┘
# [exception] Unix.Unix_error(Unix.EPERM, "bind", "")
#             Raised by primitive operation at Lwt_unix.bind in file "src/unix/lwt_unix.cppo.ml", line 1789, characters 52-74
#             Called from Lwt_io.establish_server_generic in file "src/unix/lwt_io.ml", line 1678, characters 4-52
#             Called from Lwt_io.establish_server_with_client_socket in file "src/unix/lwt_io.ml", line 1716, characters 4-88
#             Called from Helpers__Server.with_server in file "tests/helpers/server.ml", line 74, characters 11-24
#             Called from Dune__exe__Hlc_test.exec in file "tests/hlc_test.ml", line 13, characters 21-52
#             Called from Alcotest_engine__Core.Make.protect_test.(fun) in file "src/alcotest-engine/core.ml", line 180, characters 17-23
#             Called from Alcotest_engine__Monad.Identity.catch in file "src/alcotest-engine/monad.ml", line 24, characters 31-35
#             
# Logs saved to `~/.opam/default/.opam-switch/build/terminus-hlc.0.1.0/_build/default/tests/_build/_tests/Http Http-lwt-client module/call.000.output'.
#  ──────────────────────────────────────────────────────────────────────────────
# 
# Full test results in `~/.opam/default/.opam-switch/build/terminus-hlc.0.1.0/_build/default/tests/_build/_tests/Http Http-lwt-client module'.
# 4 failures! in 0.001s. 4 tests run.
```
The macOS backend is experimental so it could be an issue on our side but just in case could you investigate?

**Comment by maiste** [comment]
Sure, no problem!

**Comment by maiste** [comment]
Quick update:
I add a small test suite on my side to run with _Github Actions_ and it seems to work fine with _MacOs_ ([test suite](https://github.com/maiste/terminus/runs/7363062802?check_suite_focus=true)). However, I still need to investigate the origin of the error with the MacOs worker.

**Comment by maiste** [comment]
Sorry for the delay,
I fixed the issue with OCaml 5. I investigated a bit for the issue with MacOS, and, as I mentioned above, the execute work with the `Github Actions` runner. Maybe it is because I'm trying to bind a local port, and the CI doesn't accept it?

**Comment by mseri** [comment]
That is possible indeed, macos sandbox is more restrictive on network use

File: packages/terminus-hlc/terminus-hlc.0.1.0/opam [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  ["dune" "subst"] {dev}
+  [
+    "dune"
+    "build"
+    "-p"
+    name
+    "-j"
+    jobs
+    "@install"
+    "@runtest" {with-test}
```
Comment (mseri) [inline packages/terminus-hlc/terminus-hlc.0.1.0/opam:33]:
```suggestion
    "@runtest" {with-test & os != "macos"}
```

File: packages/terminus-cohttp/terminus-cohttp.0.1.0/opam [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  ["dune" "subst"] {dev}
+  [
+    "dune"
+    "build"
+    "-p"
+    name
+    "-j"
+    jobs
+    "@install"
+    "@runtest" {with-test}
```
Comment (mseri) [inline packages/terminus-cohttp/terminus-cohttp.0.1.0/opam:33]:
```suggestion
    "@runtest" {with-test & os != "macos"}
```

Comment (maiste) [reply packages/terminus-cohttp/terminus-cohttp.0.1.0/opam]: Thanks! :+1:

**Comment by maiste** [comment]
Thanks for your time and your help :smile:

**Pull request merged**